File: sketches/videotext/src/render/wrap.cpp
// Tags cost no columns: markup parsing already stripped them, so only
// visible codepoints count toward a row's width.

#include <string>

#include "markup.h"

namespace vt {
namespace {

// Keeps the first maxRows rows and counts the rest, so a long body never
// holds more than a screenful of rows in memory.
struct RowSink {
    int maxRows;
    std::vector<CellRow> rows;
    std::size_t count = 0;

    void push(CellRow&& row) {
        if (count < static_cast<std::size_t>(maxRows)) rows.push_back(std::move(row));
        count++;
    }
};

struct Word {
    std::size_t start;
    std::size_t length;
};

std::vector<Word> splitWords(const CellRow& cells) {
    std::vector<Word> words;
    std::size_t start = 0;
    for (std::size_t i = 0; i <= cells.size(); i++) {
        if (i == cells.size() || cells[i].codepoint == U' ') {
            if (i > start) words.push_back({start, i - start});
            start = i + 1;
        }
    }
    return words;
}

void packWords(const CellRow& cells, const std::vector<Word>& words, std::size_t columns, RowSink& sink) {
    std::size_t emitted = 0;
    CellRow currentRow;
    std::size_t currentWidth = 0;
    auto emit = [&](CellRow&& row) {
        sink.push(std::move(row));
        emitted++;
    };

    for (Word word : words) {
        // A word wider than the row is hard-broken onto rows of its own.
        while (word.length > columns) {
            if (!currentRow.empty()) {
                emit(std::move(currentRow));
                currentRow = CellRow();
                currentWidth = 0;
            }
            emit(CellRow(cells.begin() + word.start, cells.begin() + word.start + columns));
            word.start += columns;
            word.length -= columns;
        }
        if (word.length == 0) continue;

        const std::size_t neededWidth = currentWidth == 0 ? word.length : currentWidth + 1 + word.length;
        if (neededWidth > columns) {
            emit(std::move(currentRow));
            currentRow.assign(cells.begin() + word.start, cells.begin() + word.start + word.length);
            currentWidth = word.length;
        } else {
            // The space between words is untagged at a tag boundary, so a band
            // gets even padding on both sides, but keeps the tag when both
            // words share it, so "{yellow}Partly sunny{/}" is one band, not two.
            if (currentWidth > 0) {
                const Tag prevTag = currentRow.back().tag;
                currentRow.push_back({U' ', prevTag == cells[word.start].tag ? prevTag : NO_TAG});
            }
            currentRow.insert(currentRow.end(), cells.begin() + word.start, cells.begin() + word.start + word.length);
            currentWidth = neededWidth;
        }
    }
    if (!currentRow.empty() || emitted == 0) emit(std::move(currentRow));
}

}  // namespace

Wrapped wrap(const std::vector<MarkupLine>& lines, int columns, int maxRows) {
    if (columns < 1) columns = 1;
    if (maxRows < 0) maxRows = 0;
    const std::size_t cols = static_cast<std::size_t>(columns);

    RowSink sink{maxRows, {}, 0};
    std::size_t croppedFences = 0;

    for (const MarkupLine& line : lines) {
        if (line.fenced) {
            if (line.cells.size() > cols) {
                sink.push(CellRow(line.cells.begin(), line.cells.begin() + columns));
                croppedFences++;
            } else {
                sink.push(CellRow(line.cells));
            }
        } else {
            packWords(line.cells, splitWords(line.cells), cols, sink);
        }
    }

    Wrapped result;
    if (croppedFences > 0) result.warnings.push_back(std::to_string(croppedFences) + " fenced line(s) cropped");
    if (sink.count > static_cast<std::size_t>(maxRows)) {
        result.warnings.push_back(std::to_string(sink.count - maxRows) + " lines truncated");
    }
    result.rows = std::move(sink.rows);
    result.naturalRows = sink.count;
    return result;
}

}  // namespace vt

File: src/render/wrap.js
// Greedy word wrap over markup.parse's spans, plus vertical truncation to a
// row budget. Tag spans cost 0 columns on their own — only the visible text
// in each span counts toward line width, since colour markers are already
// stripped out by markup.js. Fenced lines are never wrapped, only cropped.

function flattenChars(spans) {
    const chars = [];
    for (const span of spans) {
        for (const ch of span.text) chars.push({ ch, tag: span.tag });
    }
    return chars;
}

function charsToSpans(chars) {
    const spans = [];
    let current = null;
    for (const c of chars) {
        if (current && current.tag === c.tag) {
            current.text += c.ch;
        } else {
            current = { text: c.ch, tag: c.tag };
            spans.push(current);
        }
    }
    return spans;
}

function splitWords(chars) {
    const words = [];
    let current = [];
    for (const c of chars) {
        if (c.ch === " ") {
            if (current.length > 0) {
                words.push(current);
                current = [];
            }
        } else {
            current.push(c);
        }
    }
    if (current.length > 0) words.push(current);
    return words;
}

function packWords(words, columns) {
    const rows = [];
    let currentRow = [];
    let currentWidth = 0;

    for (let word of words) {
        while (word.length > columns) {
            if (currentRow.length > 0) {
                rows.push(currentRow);
                currentRow = [];
                currentWidth = 0;
            }
            rows.push(word.slice(0, columns));
            word = word.slice(columns);
        }
        if (word.length === 0) continue;

        const neededWidth = currentWidth === 0 ? word.length : currentWidth + 1 + word.length;
        if (neededWidth > columns) {
            rows.push(currentRow);
            currentRow = [...word];
            currentWidth = word.length;
        } else {
            // The inter-word space is always untagged (finding m16): it used
            // to take the *next* word's tag, which coloured one cell of
            // padding on the left of a tagged word but not the right,
            // reading as lopsided band padding.
            if (currentWidth > 0) currentRow.push({ ch: " ", tag: null });
            currentRow.push(...word);
            currentWidth = neededWidth;
        }
    }
    if (currentRow.length > 0 || rows.length === 0) rows.push(currentRow);
    return rows;
}

export function wrap(parsedLines, columns, maxRows) {
    const warnings = [];
    const outRows = [];
    let croppedFences = 0;

    for (const line of parsedLines) {
        const chars = flattenChars(line.spans);
        if (line.fenced) {
            let rowChars = chars;
            if (rowChars.length > columns) {
                rowChars = rowChars.slice(0, columns);
                croppedFences++;
            }
            outRows.push(charsToSpans(rowChars));
        } else {
            for (const rowChars of packWords(splitWords(chars), columns)) {
                outRows.push(charsToSpans(rowChars));
            }
        }
    }

    if (croppedFences > 0) warnings.push(`${croppedFences} fenced line(s) cropped`);

    let rows = outRows;
    if (rows.length > maxRows) {
        const truncated = rows.length - maxRows;
        rows = rows.slice(0, maxRows);
        warnings.push(`${truncated} lines truncated`);
    }

    return { rows, warnings };
}
